server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -288,21 +288,30 @@ public static class Builder extends FieldMapper.Builder {
         final TextParams.Analyzers analyzers;
 
         public Builder(String name, IndexAnalyzers indexAnalyzers, boolean isSyntheticSourceEnabled) {
-            this(name, IndexVersion.current(), indexAnalyzers, isSyntheticSourceEnabled);
+            this(name, IndexVersion.current(), indexAnalyzers, isSyntheticSourceEnabled, false);
         }
 
-        public Builder(String name, IndexVersion indexCreatedVersion, IndexAnalyzers indexAnalyzers, boolean isSyntheticSourceEnabled) {
-            super(name);
+        public Builder(
+            String name,
+            IndexVersion indexCreatedVersion,
+            IndexAnalyzers indexAnalyzers,
+            boolean isSyntheticSourceEnabled,
+            boolean withinMultiField
+        ) {
+            super(name, withinMultiField);
 
             // If synthetic source is used we need to either store this field
             // to recreate the source or use keyword multi-fields for that.
             // So if there are no suitable multi-fields we will default to
             // storing the field without requiring users to explicitly set 'store'.
             //
             // If 'store' parameter was explicitly provided we'll reject the request.
+            // Note that if current builder is a multi field, then we don't need to store, given that responsibility lies with parent field
             this.store = Parameter.storeParam(
                 m -> ((TextFieldMapper) m).store,
-                () -> isSyntheticSourceEnabled && multiFieldsBuilder.hasSyntheticSourceCompatibleKeywordField() == false
+                () -> isSyntheticSourceEnabled
+                    && this.withinMultiField == false
+                    && multiFieldsBuilder.hasSyntheticSourceCompatibleKeywordField() == false
```

Comment:
This parameter is not updatable so this change should be gated by a new index version?

Reply:
True, although at the Lucene level I don't think this would fail. Fields will just stopped having stored fields at some point. It just that default of store mapping param would change after upgrade. In case of match_only_text, you wouldn't know, since store isn't a mapping param for that field type.

In any way, I can gate this behind index version.